File: apps/backend/src/modules/invoices/dto/create-invoice.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import {
  IsString,
  IsOptional,
  IsEnum,
  IsNumber,
  IsInt,
  IsDateString,
  ValidateNested,
  IsArray,
  Min,
} from 'class-validator';
import { InvoiceStatus } from '@prisma/client';

export class InvoiceItemDto {
  @ApiProperty({ example: 'Trasporto Milano - Roma', description: 'Descrizione voce' })
  @IsString()
  description: string;

  @ApiPropertyOptional({ example: 1, default: 1, description: 'Quantità' })
  @IsOptional()
  @IsInt()
  @Min(1)
  quantity?: number;

  @ApiProperty({ example: 850.00, description: 'Prezzo unitario (€)' })
  @IsNumber()
  @Min(0)
  unitPrice: number;

  @ApiProperty({ example: 850.00, description: 'Prezzo totale voce (€)' })
  @IsNumber()
  @Min(0)
  totalPrice: number;

  @ApiPropertyOptional({ description: 'ID viaggio associato' })
  @IsOptional()
  @IsString()
  tripId?: string;
}

export class CreateInvoiceDto {
  @ApiProperty({ example: 'FT-2024-001', description: 'Numero fattura' })
  @IsString()
  invoiceNumber: string;

  @ApiProperty({ description: 'ID cliente' })
  @IsString()
  clientId: string;

  @ApiProperty({ example: '2024-12-01', description: 'Data emissione' })
  @IsDateString()
  issueDate: string;

  @ApiProperty({ example: '2024-12-31', description: 'Data scadenza' })
  @IsDateString()
  dueDate: string;

  @ApiPropertyOptional({
    enum: InvoiceStatus,
    default: 'draft',
    description: 'Stato fattura',
  })
  @IsOptional()
  @IsEnum(InvoiceStatus)
  status?: InvoiceStatus;

  @ApiProperty({ type: [InvoiceItemDto], description: 'Voci fattura' })
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => InvoiceItemDto)
  items: InvoiceItemDto[];

  @ApiProperty({ example: 850.00, description: 'Imponibile (€)' })
  @IsNumber()
  @Min(0)
  subtotal: number;

  @ApiPropertyOptional({ example: 22, default: 22, description: 'Aliquota IVA (%)' })
  @IsOptional()
  @IsNumber()
  vatRate?: number;

  @ApiProperty({ example: 187.00, description: 'Importo IVA (€)' })
  @IsNumber()
  @Min(0)
  vatAmount: number;

  @ApiProperty({ example: 1037.00, description: 'Totale fattura (€)' })
  @IsNumber()
  @Min(0)
  total: number;

  @ApiPropertyOptional({ description: 'Note aggiuntive' })
  @IsOptional()
  @IsString()
  notes?: string;

  @ApiPropertyOptional({ description: 'Data pagamento' })
  @IsOptional()
  @IsDateString()
  paidDate?: string;
}

File: apps/backend/src/main/java/com/truckfleet/dto/invoice/CreateInvoiceDto.java
package com.truckfleet.dto.invoice;

import com.truckfleet.entity.enums.InvoiceStatus;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Data;

import java.math.BigDecimal;
import java.util.List;

@Data
@Schema(description = "DTO per la creazione di una fattura")
public class CreateInvoiceDto {

    @NotBlank(message = "Il numero fattura e obbligatorio")
    @Schema(example = "FT-2024-001", description = "Numero fattura")
    private String invoiceNumber;

    @NotBlank(message = "L'ID cliente e obbligatorio")
    @Schema(description = "ID del cliente")
    private String clientId;

    @NotBlank(message = "La data emissione e obbligatoria")
    @Schema(example = "2024-01-15T00:00:00", description = "Data emissione")
    private String issueDate;

    @NotBlank(message = "La data scadenza e obbligatoria")
    @Schema(example = "2024-02-15T00:00:00", description = "Data scadenza")
    private String dueDate;

    @Schema(description = "Stato della fattura")
    private InvoiceStatus status;

    @NotNull(message = "Le voci fattura sono obbligatorie")
    @Valid
    @Schema(description = "Voci della fattura")
    private List<InvoiceItemDto> items;

    @NotNull(message = "Il subtotale e obbligatorio")
    @Min(value = 0, message = "Il subtotale non puo essere negativo")
    @Schema(example = "1000.00", description = "Subtotale")
    private BigDecimal subtotal;

    @Schema(example = "22", description = "Aliquota IVA")
    private BigDecimal vatRate;

    @NotNull(message = "L'importo IVA e obbligatorio")
    @Min(value = 0, message = "L'importo IVA non puo essere negativo")
    @Schema(example = "220.00", description = "Importo IVA")
    private BigDecimal vatAmount;

    @NotNull(message = "Il totale e obbligatorio")
    @Min(value = 0, message = "Il totale non puo essere negativo")
    @Schema(example = "1220.00", description = "Totale fattura")
    private BigDecimal total;

    @Schema(description = "Note")
    private String notes;

    @Schema(example = "2024-01-20T00:00:00", description = "Data pagamento")
    private String paidDate;
}
